src/components/core/GeoChart.tsx

```diff
@@ -56,89 +77,219 @@ export class GeoChartInner extends BaseVisualization<IGeoChartInnerProps, {}> {
-            config,
-            position,
-            sizeData,
-            sizeFormat,
-            colorData,
-            colorFormat,
-            isColorLegendVisible,
-        };
-        return legendRenderer(legendProps);
+        if (sizeData) {
+            const sizeLegendOptions = {
+                sizeData,
+                sizeFormat: getFormatFromExecutionResponse(size.index, executionResponse),
+            };
+            legendProps = { ...legendProps, ...sizeLegendOptions };
+        }
+
+        if (colorData) {
+            const colorLegendOptions = {
+                colorData,
+                colorFormat: getFormatFromExecutionResponse(color.index, executionResponse),
+            };
+            legendProps = { ...legendProps, ...colorLegendOptions };
+        }
+
+        if (segment && this.state.enabledLegendItems.length) {
+            const segmentLegendOptions = {
+                segmentData: this.state.enabledLegendItems,
+                onItemClick: this.onLegendItemClick,
+            };
+            legendProps = { ...legendProps, ...segmentLegendOptions };
+        }
+
+        return legendProps;
+    }
+
+    private getChartProps(): IGeoChartRendererProps {
+        const { geoChartOptions } = this;
+        const { config, execution, afterRender } = this.props;
+
+        if (!geoChartOptions) {
+            return {
+                config,
+                execution,
+                afterRender,
+                geoData: {},
+            };
+        }
+
+        const { geoData } = geoChartOptions;
+        const segmentIndex: number = get(geoChartOptions, "geoData.segment.index");
+        let chartProps: IGeoChartRendererProps = { config, execution, afterRender, geoData };
+
+        if (segmentIndex) {
+            const selectedSegmentNames: string[] = this.state.enabledLegendItems.reduce(
+                (result: string[], item: IPushpinCategoryLegendItem): string[] => {
+                    if (item.isVisible) {
+                        result.push(item.name);
+                    }
+                    return result;
+                },
+                [],
+            );
+            chartProps = {
+                ...chartProps,
+                config: { ...config, selectedSegmentItems: selectedSegmentNames },
+            };
+        }
+
+        return chartProps;
+    }
+
+    private renderChart = (): React.ReactElement => {
+        const { chartRenderer } = this.props;
+        const chartProps: IGeoChartRendererProps = this.getChartProps();
+        return chartRenderer(chartProps);
     };
 
-    private validateData = (): void => {
-        const {
-            config: { limit = DEFAULT_DATA_POINTS_LIMIT, mdObject: { buckets = [] } = {} },
-            execution,
-            onDataTooLarge,
-        } = this.props;
+    private renderLegend = (): React.ReactElement => {
+        const { legendRenderer } = this.props;
+        const legendProps = this.getLegendProps();
```

Comment:
please add type for `legendProps` to be consistent with this line https://github.com/gooddata/gooddata-react-components/pull/1353/files#diff-401c51ce84aba39a6e5b2ba4c990ca86R271

Reply:
fixed